Services support for external workloads

jrajahalme opened the pull request:
Refactor operator's services watcher to be usable also from `clustermesh-apiserver`. Make Cilium agent sync services from kvstore in `join-cluster` mode. Use `cilium service` to locate `kube-dns` and overwrite VM's `/etc/resolv.conf` with it.

None of this should affect any other functionality, so this should be safe to merge.

In `pkg/k8s/service_cache.go`, function `correlateEndpoints()` a check for `option.Config.ClusterName` is removed. This is required as cluster services synced from the k8s cluster will have the same cluster name as the external endpoint has, as it is conceptually in the same cluster as the k8s cluster it joined. Clustermesh call paths have the same check so it is still not possible for a clustermesh to re-introduce backends for it's own cluster via external endpoints.

'experimental-install.yaml' is changed to not configure support for external workloads, so an explicit helm option is required. Also, a new helm value `clustermesh.useAPIServer` is added and `clustermesh-apiserver` is only deployed if that or `externalWorkloads.enabled` is `true`, both default to `false`. This avoids surprising existing users of clustermesh with a new deployment.

Scripts are added to `contrib/k8s` to extract TLS certs as well as to install Cilium agent onto a VM with docker. Service access requires `--enable-host-reachable-services`, which does not work on older kernels. We could soften this requirement by not requiring that service access from VM is usable. Currently kube-dns is accessed via it's ClusterIP service, though.

TODO:
- [ ] CI test(s)
- [x] GSG

For later PRs:
- [ ] Handle headless services (== no frontends)
- [ ] test with IPv6 instead of IPv4
- [ ] test with direct routing modes, remove the need for IPAM setup in this case
- [ ] check if tunneling can be done without any IPAM (i.e., w/o creating `cilium_host` device)
- [ ] pass policy from cluster to the VMs
- [ ] enforce policy in VMs

Fixes: #13666

Test setup consists of two VMs, one running with k8s (`k8s1`), one running without k8s (`runtime`). Setup steps:

1. `cd cilium/test`
2. Start `k8s1`:
`VM_CPUS=6 VM_MEMORY=5120 K8S_NODES=1 ./vagrant-local-start.sh`
 - This may take a long time for the 1st run, as a vagrant box is packaged (into `test/.vagrant/k8s-box-package.box`). Subsequent runs will be faster. This packaging removes the need to pull test images for later runs of `./vagrant-local-start.sh`.
3. `vagrant ssh k8s1-1.19`
4. `cd ~/go/src/github.com/cilium/cilium`
5. `DOCKER_BUILDKIT=1 make docker-cilium-image docker-operator-generic-image docker-clustermesh-apiserver-image`
6. `helm install cilium install/kubernetes/cilium --namespace kube-system --set externalWorkloads.enabled=true --set image.pullPolicy=IfNotPresent --set operator.image.pullPolicy=IfNotPresent --set preflight.image.pullPolicy=IfNotPresent --set clustermesh.apiserver.image.pullPolicy=IfNotPresent --set debug.enabled=true`
7. Apply the CiliumExternalWorkload CRD for VM named `runtime`: `kubectl apply -f clustermesh-apiserver/test/runtime.yaml`
8. Extract the TLS client keys to be used by the VM that is started up in the next step:
`contrib/k8s/extract-external-workload-certs.sh`
9. In another terminal, change to `test` directory in your Cilium repo and start the `runtime` VM:
`PROVISION_EXTERNAL_WORKLOAD=1 PRELOAD_VM=false KERNEL=419 VM_CPUS=6 VM_MEMORY=4096 ./vagrant-local-start-runtime.sh`
10. `vagrant ssh runtime`
11. Check that etcd has connected successfully, observe the first line of:
`cilium status`
You should see something like:
`etcd: 1/1 connected, lease-ID=7c02748328e75f57, lock lease-ID=7c02748328e75f59, has-quorum=true: 192.168.36.11:32379 - 3.4.13 (Leader)`
12. Check that cluster DNS works: `nslookup clustermesh-apiserver.kube-system.svc.cluster.local`
13. Ping the backend IP of `clustermesh-apiserver` service:
`ping $(cilium service list get -o jsonpath='{[?(@.spec.flags.name=="clustermesh-apiserver")].spec.backend-addresses[0].ip}')`
14. The ping should keep running also when the following CCNP is applied in `k8s1` (e.g., `kubectl apply -n kube-system -f clustermesh-apiserver/test/testccnp.yaml`):
```
apiVersion: cilium.io/v2
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: test-ccnp
  namespace: kube-system
spec:
  endpointSelector:
    matchLabels:
      k8s-app: clustermesh-apiserver
  ingress:
  - fromEndpoints:
    - matchLabels:
        io.kubernetes.pod.name: runtime
  - toPorts:
    - ports:
      - port: "2379"
        protocol: TCP
```
15. The ping should stop if you delete these lines from the policy and re-apply:
```
  - fromEndpoints:
    - matchLabels:
        io.kubernetes.pod.name: runtime
```
16. `kubectl get cep` should show an entry for `runtime` with `IDENTITY ID` and `IPV4` set.
17. `kubectl get cew` should show an entry for `runtime` with matching `CILIUM ID` and `IP`.

jrajahalme committed:
clustermesh-apiserver: Watch for endpoints and endpointslices

Re-use operator's service watcher to add endpoints and endpointslices
watchers so that services being inserted into kvstore have the
backends.

Make daemon watch for services in the kvstore in join-cluster
mode. Add code to create Cilium services from kvstore services in
join-cluster mode.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
operator: Refactor services watcher to watchers package

Move k8s_service_sync.go to watchers so that it can be re-used by
importing.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Revert "install/kubernetes: Enable external workloads in experimental-install.yaml"

This reverts commit 2345e536154c8a6492df880e2e2b89e2215b83c6.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
helm: Do not deploy clustermesh-apiserver for clustermesh by default

Require explicit setting of helm value 'clustermesh.useAPIServer=true'
to deploy clustermesh-apiserver when using clustermesh.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

christarazi commented on pkg/k8s/service.go:
```diff
@@ -483,6 +483,89 @@ func NewClusterService(id ServiceID, k8sService *Service, k8sEndpoints *Endpoint
+		comparator.MapStringEquals(s.Selector, svc.Selector) &&
+		s.SessionAffinity == false &&
+		s.SessionAffinityTimeoutSec == 0 &&
+		s.Type == loadbalancer.SVCTypeClusterIP {
```
I think we should move this to a helper function. It's quite a long list of conditionals that could use some explanation via moving to a helper.

jrajahalme replied:
The whole function is an "Equals" function. I'll add a comment to refer to the parse function right above for the detail.

aanm replied:
Can we have this function auto generated please?

jrajahalme replied:
This can not be auto-generated, as it compares two different types. We need this to avoid re-creating each service when the backends change, as the ClusterService in kvstore represents both the service and the backends in the same value.

christarazi commented on pkg/service/store/store.go:
```diff
@@ -133,3 +144,59 @@ func NewClusterService(name, namespace string) ClusterService {
+}
+
+// Controller for syncing services from the kvstore
+func JoinClusterServices(merger ServiceMerger) {
```
Nit: godoc doesn't match up with the function name. Also could you expand a bit more on this one? The current comment doesn't give much info to me.

jrajahalme replied:
Will do.

christarazi commented on operator/watchers/k8s_service_sync.go:
```diff
@@ -48,6 +48,7 @@ var (
 	// k8s.
 	k8sSvcCacheSynced = make(chan struct{})
 	kvs               *store.SharedStore
+	sharedOnly        bool
```
Could you describe why this is needed and what it stands for?

jrajahalme replied:
Operator use (the original use) syncs only shared services. Clustermesh-apiserver needs to sync all services so that the VMs get access to them.

christarazi replied:
Could you add that to src? :)

jrajahalme replied:
Done.

christarazi reviewed (changes requested):
A few minor comments on the Go code, all other changes LGTM.

nebril commented on contrib/k8s/install-external-workload.sh:
```diff
@@ -0,0 +1,131 @@
+# Wait for kube-dns service to become available
+kubedns=""
+for ((i = 0 ; i < 24; i++)); do
+    kubedns=`cilium service list get -o jsonpath='{[?(@.spec.frontend-address.port==53)].spec.frontend-address.ip}'`
```
There are couple other shellcheck issues in this file (lack of double quotes around variables), but I think we should avoid using backtics and use `$(...)`. (http://mywiki.wooledge.org/BashFAQ/082)

jrajahalme replied:
Thanks, will change :-)

jrajahalme replied:
Done.

jrajahalme committed:
contrib/k8s: Add scripts for provisioning VMs

Add script 'extract-external-workload-certs.sh' to extract external
workload certs from k8s cluster to be used by a VM. Certs are placed
in the current directory, which can be overridden by defining
TLS_PATH.

Add script 'install-external-workload.sh' to run Cilium in a docker
container in a VM. CLUSTER_ADDR must be defined as the IP in the k8s
cluster at which 'clustermesh-apiserver' is available. TLS certs need
to exist in the current directory, or in dir TLS_PATH. Default Cilium
image can be overridden with CILIUM_IMAGE. VM's IP address can be
selected by defining 'NODE_IP'. Script waits for Cilium and kube-dns
service to become available and overrides the VM DNS config to use
kube-dns as the nameserver.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on pkg/k8s/service_cache.go:
```diff
@@ -397,7 +397,7 @@ func (s *ServiceCache) UniqueServiceFrontends() FrontendList {
 // returns a boolean that indicates whether the service is ready to be plumbed,
 // this is true if:
-// IF If ta local endpoints resource is present. Regardless whether the
+// IF If a local endpoints resource is present. Regardless whether the
```
nit, there's three ifs in a row between these two lines :thinking:

jrajahalme replied:
fixed.

joestringer commented on pkg/k8s/service_cache.go:
```diff
@@ -448,18 +444,21 @@ func (s *ServiceCache) correlateEndpoints(id ServiceID) (*Endpoints, bool) {
+	s.mergeServiceUpdate(service, nil, swg)
+}
+
+func (s *ServiceCache) mergeServiceUpdate(service *serviceStore.ClusterService, oldService *Service, swg *lock.StoppableWaitGroup) {
```
nit:
```suggestion
func (s *ServiceCache) mergeServiceUpdateLocked(service *serviceStore.ClusterService, oldService *Service, swg *lock.StoppableWaitGroup) {
```

jrajahalme replied:
Applied.

joestringer commented on pkg/k8s/service.go:
```diff
@@ -483,6 +483,105 @@ func NewClusterService(id ServiceID, k8sService *Service, k8sEndpoints *Endpoint
 	return svc
 }
 
+// ParseClusterService parses a ClusterService and returns a Service
```
```suggestion
// ParseClusterService parses a ClusterService and returns a Service.
```

joestringer reviewed (approved):
I didn't do a thorough review, but I glanced over all the changes. Only thing I could tell could plausibly impact operator is the refactoring and any other changes around the service watchers (like handling events for the current cluster as well as remote cluster).

LGTM. Minor nits.
